src/NHibernate/Mapping/ByCode/Impl/CustomizersImpl/PropertyContainerCustomizer.cs

```diff
@@ -94,14 +94,59 @@ protected void RegisterComponentMapping<TComponent>(Action<IComponentMapper<TCom
 			}
 		}
 
+		/// <summary>
+		/// Maps a non-generic dictionary property as a dynamic component.
+		/// </summary>
+		/// <param name="property">The property to map.</param>
+		/// <param name="dynamicComponentTemplate">The template for the component. It should either be a (usually
+		/// anonymous) type having the same properties than the component, or an
+		/// <c>IDictionary&lt;string, System.Type&gt;</c> of property names with their type.</param>
+		/// <param name="mapping">The mapping of the component.</param>
+		/// <typeparam name="TComponent">The type of the template.</typeparam>
 		public void Component<TComponent>(Expression<Func<TEntity, IDictionary>> property, TComponent dynamicComponentTemplate, Action<IDynamicComponentMapper<TComponent>> mapping)
 		{
-			RegisterDynamicComponentMapping(property, mapping);
+			if (dynamicComponentTemplate is IDictionary<string, System.Type> dicTemplate && mapping is Action<IDynamicComponentMapper<object>> objectMapping)
```

Comment:
Is it still valid check? It seems separate method needs to be used for dictionary.

Reply:
>Is it still valid check?

Yes

> It seems separate method needs to be used for dictionary.

Unfortunately this is not really possible because the generic instance overload takes precedence over extension methods.

Reply:
Hm... But `TComponent` here is  `Dictionary<string, Type>`.
I don't get how `Action<IDynamicComponentMapper<Dictionary<string, Type>>>` can be casted to `Action<IDynamicComponentMapper<object>> objectMapping`

Reply:
It's easy. `typeof(TComponent) == typeof(object)`.

Reply:
But it means user needs to explicitly cast to `object` to make it work. Not obvious at all.  This I'm pretty sure will lead to many confusions. 
Wouldn't it be better to simply introduce separate method for dictionary? Something like `ComponentForDic`

Reply:
> Not obvious at all. This I'm pretty sure will lead to many confusions.

This is what I've meant [here](https://github.com/nhibernate/nhibernate-core/pull/1941#discussion_r268542709).

Reply:
So  solution would be to add separate method. This avoids castings here and makes new method discoverable and user doesn't need to know that to make it work explicit cast to object is required. 

What disadvantages do you see?

Reply:
I've solved this by just checking for dictionary with any type of mapper. So, it's not a problem anymore.

Reply:
Cool